dhis-2/dhis-api/src/main/java/org/hisp/dhis/dataelement/DataElement.java

```diff
@@ -172,6 +172,25 @@ public DataElement( String name )
         this.name = name;
     }
 
+    public DataElement( DataElement de )
+    {
+        super( de );
```

Comment:
Obviously having "copy constructors" is not ideal, as adds "boilerplate" and is prone to future issues when new properties are added to the data model. Are you able to think of another solution which does not require copying of the domain model @jimgrace ?

Reply:
For why copies are needed, please see my Feb 24 comment above to @jbee. Options include:

1. Copy constructors
Cons:
o Boilerplate code to maintain
Pros:
o Fast
o Minimal special-case code for .periodOffset in DefaultDimensionService
o The copy constructors are at least easy to review and are near the object property definitions

2. Case-specific copy helpers
Cons:
o May not be much simpler than the copy constructors
o May break more easily if, for example, a new (or old) DataElement field is newly used in Analytics that is not in the copy helper. Not near the property definitions.
Pros:
o Fast
o Minimal special-case code for .periodOffset in DefaultDimensionService

3. SerializationUtils.clone, used in DefaultDimensionService.getItemObjectMap only when there is a .periodOffset present (or other query modifier in the future like data element aggregation type override).
Cons:
o Some performance hit, but only when .periodOffset is used
o Some more special-case .periodOffset code in DefaultDimensionService
Pros:
o Simple code, no boilerplate!

4. SerializationUtils.clone, used always in DefaultDimensionService.getItemObjectMap in place of the copy constructors
Cons:
o Some performance hit always
Pros:
o Simple code, no boilerplate!
o Minimal special-case code for .periodOffset in DefaultDimensionService

I think option 4 looks very attractive if the performance hit is negligible. I am willing to do some timing testing to determine this, if you like. We could also consider option 3 as a compromise if the performance hit is not negligible but not too bad.